File: CS/Advance/queue/minWindowArrK.cpp
#include <cmath>
#include <deque>
#include <iostream>
#include <iterator>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

vector<int> minWindowArrK(vector<int> intArr, int k) {
    if (k == 1) return intArr;

    deque<int> deque;
    vector<int> answer;

    for (int i = 0; i < k; i++) {
        while (!deque.empty() && intArr[deque.back()] >= intArr[i]) {
            deque.pop_back();
        }
        deque.push_back(i);
    }

    for (int i = k; i < intArr.size(); i++) {
        answer.push_back(intArr[deque.front()]);

        while (!deque.empty() && deque.front() <= i - k) deque.pop_front();

        while (!deque.empty() && intArr[deque.back()] >= intArr[i])
            deque.pop_back();

        deque.push_back(i);
    }

    answer.push_back(intArr[deque.front()]);

    return answer;
}

File: CS/Advance/queue/minWindowArrK.py
from collections import deque


def minWindowArrK(intArr, k) -> list:
    dq = deque()
    answer = []
    for i in range(0, k):
        while len(dq) != 0 and intArr[i] < intArr[dq[-1]]:
            dq.pop()
        dq.append(i)

    for i in range(k, len(intArr)):
        answer.append(intArr[dq[0]])

        while len(dq) != 0 and dq[0] <= i - k:
            dq.popleft()

        while len(dq) != 0 and intArr[i] < intArr[dq[-1]]:
            dq.pop()

        dq.append(i)


    answer.append(intArr[dq.popleft()])
    return answer


print(minWindowArrK([24,5,67,60,24,64,23,536,345],3))
